src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs

```diff
@@ -0,0 +1,260 @@
+                {
+                    return attr;
+                }
+            }
+            return new CustomAttributeHandle();
+        }
+
+        private static string GetTypeGuid(MetadataReader reader, TypeDefinition type)
+        {
+            foreach (CustomAttributeHandle attr in type.GetCustomAttributes())
+            {
+                CustomAttribute attribute = reader.GetCustomAttribute(attr);
+                MemberReference attributeConstructor = reader.GetMemberReference((MemberReferenceHandle)attribute.Constructor);
+                TypeReference attributeType = reader.GetTypeReference((TypeReferenceHandle)attributeConstructor.Parent);
+                if (reader.GetString(attributeType.Namespace) == "System.Runtime.InteropServices" && reader.GetString(attributeType.Name) == "GuidAttribute")
+                {
+                    CustomAttributeValue<KnownTypes> data = attribute.DecodeValue(new TypeResolver());
+                    return (string)data.FixedArguments[0].Value;
+                }
+            }
+            throw new BuildErrorException(Strings.ClsidMapExportedTypesRequireExplicitGuid, GetTypeName(reader, type));
+        }
+
+        private enum KnownTypes
+        {
+            Bool,
+            String,
+            SystemType,
+            Unknown
+        }
+
+        private class TypeResolver : ICustomAttributeTypeProvider<KnownTypes>
+        {
+            public KnownTypes GetPrimitiveType(PrimitiveTypeCode typeCode)
+            {
+                switch (typeCode)
+                {
+                    case PrimitiveTypeCode.Boolean:
+                        return KnownTypes.Bool;
+                    case PrimitiveTypeCode.String:
+                        return KnownTypes.String;
+                    default:
+                        return KnownTypes.Unknown;
+                }
+            }
+
+            public KnownTypes GetSystemType()
+            {
+                return KnownTypes.SystemType;
+            }
+
+            public KnownTypes GetSZArrayType(KnownTypes elementType)
+            {
+                return KnownTypes.Unknown;
+            }
+
+            public KnownTypes GetTypeFromDefinition(MetadataReader reader, TypeDefinitionHandle handle, byte rawTypeKind)
+            {
+                return KnownTypes.Unknown;
+            }
+
+            public KnownTypes GetTypeFromReference(MetadataReader reader, TypeReferenceHandle handle, byte rawTypeKind)
+            {
+                return KnownTypes.Unknown;
+            }
+
+            public KnownTypes GetTypeFromSerializedName(string name)
+            {
+                return KnownTypes.Unknown;
+            }
+
+            public PrimitiveTypeCode GetUnderlyingEnumType(KnownTypes type)
+            {
+                return PrimitiveTypeCode.Int32;
```

Comment:
I think it would be better to throw BadImageFormatException here with a comment that there are no enum arguments to the attributes we care about. If ever we did encounter an enum argument to these attributes,  the stack will be more obvious if we fail now rather than reading the wrong number of bytes IMHO.

Side rant: The metadata format for attributes is really silly here in that it requires you to resolve an enum type to its definition to decide how many bytes to read. I wrote the custom attribute decoder and it pained me to have to have this method.